File: leetcode/counting-bits.cpp
#include <iostream>
#include <cassert>

#include <vector>
#include <stack>

using namespace std;

/**
https://leetcode.com/problems/counting-bits

// Soluiton1. popcount bitmanipulation (pop the least 1)
// time o(nk) k: number of bit in x
// space o(n) ans
ex) 3
11 -> 10 -> 01 -> 0 = count(3)

int count
for (count=0; n!=0; count++)
    n &= n-1;
    
// Soluiton2. bitmanipulation f(x) = f(x & (x-1)) + 1 with bp
time: o(n)
space: o(n) ans
ex) 3
11 = 10(which is 11 & 10) + 1

*/

class Solution {
public:
    int popcount(int num) {
        int count=0;
        for (;num!=0;count++)
            num &= num-1;
        return count;
    }
    
    vector<int> countBits_popcount(int num) {
        vector<int> ans;
        for (int i=0; i<=num; i++)
            ans.push_back(popcount(i));
        return ans;
    }
    
    vector<int> countBits_dp(int num) {
        vector<int> ans(num+1);
        for (int i=1; i<=num; i++)
            ans[i] = ans[i & (i-1)] + 1;
        return ans;
    }
    
    vector<int> countBits(int num) {
        //return countBits_popcount(num);
        return countBits_dp(num);
    }
};

int main()
{
    auto expected1 = vector<int> {0, 1, 1};
    assert(Solution().countBits(2)
        == (expected1));
    auto expected2 = vector<int> {0, 1, 1, 2, 1, 2};
    assert(Solution().countBits(5)
        == (expected2));

    return 0;
}

File: leetcode/leetcode_py/counting-bits.py
'''
https://leetcode.com/problems/counting-bits

1. Bit manipulation
time: o(n)
space: o(1) n:answer
res[i] = res[i >> 1] + (i & 1)
ex) 
0: 0
1: 0 + 1 = 1
2: 1(1) + 0 = 1
3: 1(1) + 1 = 2
4: 1(2) + 0 = 1
5: 1(2) + 1 = 2
6: 2(3) + 0 = 2
7: 2(3) + 1 = 3
8: 1(4) + 0 = 1
9: 1(4) + 1 = 2
'''
class Solution:
    def countBits(self, n: int) -> List[int]:
        res = [0] * (n+1)
        for i in range(1, n+1):
            res[i] = res[i >> 1] + (i & 1)
        return res
        